invenio_rdm_records/services/customizations.py

```diff
@@ -41,57 +41,133 @@ def customize(cls, opts):
-                if attrs['pids_providers'][scheme]['default'] is None:
-                    attrs['pids_providers'][scheme]['default'] = name
-
-        # Create the config class
-        return _make_cls(cls, attrs)
+            (This transformation would be unnecesary if the pid_providers were
+            in the right shape already).
+            """
+            provider_dict = {"default": None}
+
+            for name in pid_config.get("providers", []):
+                # This may throw a KeyError which is a sign that the config
+                # is wrong.
+                provider_dict[name] = pid_providers[name]
+                provider_dict['default'] = provider_dict['default'] or name
+
+            return provider_dict
+
+        pids = current_app.config.get("RDM_PERSISTENT_IDENTIFIERS", {})
+        providers = {
+            p.name: p for p in
+            current_app.config.get("RDM_PERSISTENT_IDENTIFIER_PROVIDERS", [])
+        }
+        doi_enabled = current_app.config.get("DATACITE_ENABLED", False)
+
+        return {
+            scheme: get_provider_dict(conf, providers)
+            for scheme, conf in pids.items()
+            if scheme != 'doi' or doi_enabled
+        }
+
+
+class FromConfigRequiredPIDs:
+    """Data descriptor for required pids configuration."""
+
+    def __get__(self, obj, objtype=None):
+        """Return required pids (descriptor protocol)."""
+        pids = current_app.config.get("RDM_PERSISTENT_IDENTIFIERS", {})
+        doi_enabled = current_app.config.get("DATACITE_ENABLED", False)
+
+        pids = {
+            scheme: conf for (scheme, conf) in pids.items()
+            if scheme != 'doi' or doi_enabled
+        }
+        return [
+            scheme for (scheme, conf) in pids.items()
+            if conf.get("required", False)
+        ]
+
+
+class FromConfigSearchOptions:
+    """Data descriptor for search options configuration."""
+
+    def __init__(self, config_key, default=None, search_option_cls=None):
+        """Constructor for data descriptor."""
+        self.config_key = config_key
+        self.default = default or {}
+        self.search_option_cls = search_option_cls
+
+    def __get__(self, obj, objtype=None):
+        """Return value that was grafted on obj (descriptor protocol)."""
+        search_opts = current_app.config.get(self.config_key, self.default)
+        sort_opts = current_app.config.get('RDM_SORT_OPTIONS')
+        facet_opts = current_app.config.get('RDM_FACETS')
+
+        search_config = SearchConfig(
+            search_opts,
+            sort=sort_opts,
+            facets=facet_opts,
+        )
+
+        return self.search_option_cls.customize(search_config)
+
+
+class ConfiguratorMixin:
+    """Shared customization for requests service config."""
+
+    @classmethod
+    def build(cls):
+        """Build the config object."""
+        return type(f"Custom{cls.__name__}", (cls,), {})
```

Comment:
question: what does this do?
aren't the config properties now calculated lazily when they're accessed via the descriptors, removing the necessity for an eager build?

Reply:
Hmmm if I recall, I was just keeping the additional layer that was already there here. Since we were creating this class before via `_make_cls` I kept it around. I'll make a ticket to investigate if this can be removed or if we somehow needed this additional layer for some reason: https://github.com/inveniosoftware/invenio-rdm-records/issues/872